ompi/mca/osc/ucx/osc_ucx_component.c

```diff
@@ -0,0 +1,565 @@
+        if (UCS_OK != status) {
+            goto error;
+        }
+
+        /* query UCP worker attributes */
+        worker_attr.field_mask = UCP_WORKER_ATTR_FIELD_THREAD_MODE;
+        status = ucp_worker_query(mca_osc_ucx_component.ucp_worker, &worker_attr);
+        if (UCS_OK != status) {
+            goto error;
+        }
+
+        if (mca_osc_ucx_component.enable_mpi_threads == true &&
+            worker_attr.thread_mode != UCS_THREAD_MODE_MULTI) {
+            goto error;
+        }
+    }
+
+    /* create module structure */
+    module = (ompi_osc_ucx_module_t *)calloc(1, sizeof(ompi_osc_ucx_module_t));
+    if (module == NULL) {
+        return OMPI_ERR_TEMP_OUT_OF_RESOURCE;
+    }
+
+    /* fill in the function pointer part */
+    memcpy(module, &ompi_osc_ucx_module_template, sizeof(ompi_osc_base_module_t));
+
+    ret = ompi_comm_dup(comm, &module->comm);
+    if (ret != OMPI_SUCCESS) {
+        goto error;
+    }
+
+    asprintf(&name, "ucx window %d", ompi_comm_get_cid(module->comm));
+    ompi_win_set_name(win, name);
+    free(name);
+
+    /* share everyone's displacement units. Only do an allgather if
+       strictly necessary, since it requires O(p) state. */
+    temp = disp_unit;
+    ret = module->comm->c_coll->coll_bcast(&temp, 1, MPI_INT, 0,
+                                           module->comm,
+                                           module->comm->c_coll->coll_bcast_module);
+    if (OMPI_SUCCESS != ret) {
+        goto error;
+    }
+
+    temp = (temp == disp_unit) ? 1 : 0; /* judge if everyone has the same disp unit size */
+
+    ret = module->comm->c_coll->coll_allreduce(MPI_IN_PLACE, &temp, 1, MPI_INT, MPI_LAND,
+                                               module->comm,
+                                               module->comm->c_coll->coll_allreduce_module);
+    if (OMPI_SUCCESS != ret) {
+        goto error;
+    }
+
+    if (temp == 1) { /* everyone has the same disp_unit, we do not need O(p) space */
+        module->disp_unit = disp_unit;
+    } else { /* different disp_unit sizes, allocate O(p) space to store them */
+        module->disp_unit = -1;
+        module->disp_units = calloc(comm_size, sizeof(int));
+        if (module->disp_units == NULL) {
+            goto error;
+        }
+
+        ret = module->comm->c_coll->coll_allgather(&disp_unit, 1, MPI_INT,
+                                                   module->disp_units, 1, MPI_INT,
+                                                   module->comm,
+                                                   module->comm->c_coll->coll_allgather_module);
+        if (OMPI_SUCCESS != ret) {
+            goto error;
+        }
+    }
+
+    /* exchange endpoints if necessary */
+    for (i = 0; i < comm_size; i++) {
+        if (OSC_UCX_GET_EP(module->comm, i) == NULL) {
+            collect_eps = true;
+            break;
+        }
+    }
+
+    if (collect_eps) {
```

Comment:
Can it be the case that collect_eps will be different on different ranks? If yes, then allgather will be called by not all ranks from the win communicator.